internal/helm/release/manager.go

```diff
@@ -363,3 +366,59 @@ func (m manager) UninstallRelease(ctx context.Context, opts ...UninstallOption)
 	}
 	return uninstallResponse.Release, err
 }
+
+// CleanupRelease deletes resources if they are not deleted already.
+// Return true if all the resources are deleted, false otherwise.
+func (m manager) CleanupRelease(ctx context.Context, manifest string) (bool, error) {
+	dc, err := m.actionConfig.RESTClientGetter.ToDiscoveryClient()
+	if err != nil {
+		return false, fmt.Errorf("failed to get Kubernetes discovery client: %w", err)
+	}
+	apiVersions, err := action.GetVersionSet(dc)
+	if err != nil && !discovery.IsGroupDiscoveryFailedError(err) {
+		return false, fmt.Errorf("failed to get apiVersions from Kubernetes: %w", err)
+	}
+	manifests := releaseutil.SplitManifests(manifest)
+	_, files, err := releaseutil.SortManifests(manifests, apiVersions, releaseutil.UninstallOrder)
+	if err != nil {
+		return false, fmt.Errorf("failed to sort manifests: %w", err)
+	}
+	// do not delete resources that are annotated with the Helm resource policy 'keep'
+	_, filesToDelete := manifestutil.FilterManifestsToKeep(files)
+	var builder strings.Builder
+	for _, file := range filesToDelete {
+		builder.WriteString("\n---\n" + file.Content)
+	}
+	resources, err := m.kubeClient.Build(strings.NewReader(builder.String()), false)
+	if err != nil {
+		return false, fmt.Errorf("failed to build resources from manifests: %w", err)
+	}
+	if resources == nil || len(resources) <= 0 {
+		return true, nil
+	}
+	for _, resource := range resources {
+		err = resource.Get()
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				continue // resource is already delete, check the next one.
+			}
+			return false, fmt.Errorf("failed to get resource: %w", err)
+		}
+		// found at least one resource that is not deleted so just delete everything again.
+		_, errs := m.kubeClient.Delete(resources)
+		if errs != nil {
+			return false, fmt.Errorf("failed to delete resources: %s", joinErrors(errs))
+		}
+		return false, nil
```

Comment:
This return appears to short-circuit the loop so only one resource is ever deleted.
```suggestion
```

Reply:
Thank you for the review. Could you please take a second look around this code block?
```
		// found at least one resource that is not deleted so just delete everything again.
		_, errs := m.kubeClient.Delete(resources)
		if errs != nil {
			return false, fmt.Errorf("failed to delete resources: %s", joinErrors(errs))
		}
		return false, nil
```
The helm kube client has a function that delete all the resources given (skips the ones that are deleted or not found). So rather than just delete one resource, I decided to just delete all of them again . Do you think this is fine?

Reply:
I see. Why even bother with the loop then?

Reply:
I was thinking calling the async delete resources might not guarantee the resources will be deleted (even if there is no error return, maybe it takes a while before the delete actually finish processing) . The loop is to check if the resources are actually deleted or not.

```
	for _, resource := range resources {
		err = resource.Get()
		if err != nil {
			if apierrors.IsNotFound(err) {
				continue // resource is already delete, check the next one.
			}
```
once it figures out all the resources are actually gone then return true to indicate everything is actually deleted.

Reply:
Ok, in that case I think your construction does work.